aws/hll/aws-s3-transfer-manager/src/io/error.rs

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use std::error::Error as StdError;
+use std::fmt;
+use std::fmt::Formatter;
+use std::io::{Error as StdIoError, ErrorKind as StdIoErrorKind};
+use tokio::task::JoinError;
+
+#[derive(Debug)]
+pub(crate) enum ErrorKind {
+    UpperBoundSizeHintRequired,
+    OffsetGreaterThanFileSize,
+    TaskFailed(JoinError),
+    IOError(StdIoError),
+}
+
+/// An I/O related error occurred
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+}
+
+impl Error {
+    pub(crate) fn upper_bound_size_hint_required() -> Error {
+        ErrorKind::UpperBoundSizeHintRequired.into()
+    }
+}
+impl From<ErrorKind> for Error {
+    fn from(kind: ErrorKind) -> Self {
+        Self { kind }
+    }
+}
+
+impl From<StdIoError> for Error {
+    fn from(err: StdIoError) -> Self {
+        ErrorKind::IOError(err).into()
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match &self.kind {
+            ErrorKind::UpperBoundSizeHintRequired => write!(
+                f,
+                "size hint upper bound (SizeHint::upper) is required but was None"
+            ),
+            ErrorKind::OffsetGreaterThanFileSize => write!(
+                f,
+                "offset must be less than or equal to file size but was greater than"
+            ),
+            ErrorKind::IOError(_) => write!(f, "I/O error"),
+            ErrorKind::TaskFailed(_) => write!(f, "internal task failed to complete"),
```

Comment:
Will tasks have specific failure modes in the future? We should consider using language similar to the `IoError` ones, i.e. more generic messaging like:
```suggestion
            ErrorKind::TaskError(_) => write!(f, "task error"),
```

Reply:
Or maybe
```suggestion
            ErrorKind::JoinError(_) => write!(f, "join error"),
```

Reply:
I don't know given we are currently using `tokio` but we may want to be more runtime agnostic in the future. Hard to say what it should look like.